### PR TITLE
Support protocol and encoding client options based on content-type in Studio Agent

### PR DESCRIPTION
**Pull request by doriable**
Currently we are defaulting to the gRPC protocol in Studio Agent,
however, we are able to support `application/json` and `application/proto`
with the Connect protocol. This just refactors the client options so that
we are setting protocol and codec options based on the content-type.

### CODE REVIEW

File: private/bufpkg/bufstudioagent/plain_post_handler.go [hunk 1]
```diff
@@ -222,6 +203,43 @@ func (i *plainPostHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 }
 
+func connectClientOptionsFromContentType(contentType string, requestMsgSize int) ([]connect.ClientOption, error) {
+	switch contentType {
+	case "application/grpc", "application/grpc+proto":
+		return []connect.ClientOption{
+			connect.WithGRPC(),
+			connect.WithCodec(&bufferCodec{name: "proto"}),
+		}, nil
+	case "application/grpc-json":
```
Comment (bufdev) [inline private/bufpkg/bufstudioagent/plain_post_handler.go:213]: I've never seen this before, shouldn't this be `application/grpc+json`?

Comment (doriable) [reply private/bufpkg/bufstudioagent/plain_post_handler.go:213]: Uh, yes, it should be 😬

**Comment by doriable** [comment]
Removed the empty content-type case with an empty message, since defaulting to gRPC there seems like it would be hard to debug. This way, we should always expect client to provide content-type, which is reasonable since we control the client-side interaction via Studio.

File: private/bufpkg/bufstudioagent/plain_post_handler.go [hunk 2]
```diff
@@ -222,6 +203,31 @@ func (i *plainPostHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	})
 }
 
+func connectClientOptionsFromContentType(contentType string, requestMsgSize int) ([]connect.ClientOption, error) {
```
Comment (robbertvanginkel) [inline private/bufpkg/bufstudioagent/plain_post_handler.go:206]: No need for `requestMsgSize int` anymore.

File: private/bufpkg/bufstudioagent/bufstudioagent_test.go [hunk 3]
```diff
@@ -68,34 +68,65 @@ func testPlainPostHandler(t *testing.T, upstreamServer *httptest.Server) {
...
+		assert.Equal(t, http.StatusOK, response.StatusCode)
+		assert.Equal(t, "https://example.buf.build", response.Header.Get("Access-Control-Allow-Origin"))
+		responseBytes, err := io.ReadAll(response.Body)
+		assert.NoError(t, err)
+		invokeResponse := &studiov1alpha1.InvokeResponse{}
+		protoUnmarshalBase64(t, responseBytes, invokeResponse)
+		upstreamResponseHeaders := make(http.Header)
+		addProtoHeadersToGoHeader(invokeResponse.Headers, upstreamResponseHeaders)
+		addProtoHeadersToGoHeader(invokeResponse.Trailers, upstreamResponseHeaders)
+		assert.Equal(t, []byte("echo: echothis"), invokeResponse.Body)
```
Comment (robbertvanginkel) [inline private/bufpkg/bufstudioagent/bufstudioagent_test.go:128]: Could add something like `assert.Equal(t, "", upstreamResponseHeaders.Get("grpc-status"))` to validate the response wasnt a grpc response?

File: private/bufpkg/bufstudioagent/plain_post_handler.go (hunk 2 above)
Comment (doriable) [reply private/bufpkg/bufstudioagent/plain_post_handler.go:206]: 😬 🤦🏼‍♀️

File: private/bufpkg/bufstudioagent/bufstudioagent_test.go (hunk 3 above)
Comment (doriable) [reply private/bufpkg/bufstudioagent/bufstudioagent_test.go:129]: Ah yeh, this is a good idea -- added some new assertions.

**Review by robbertvanginkel — APPROVED** [review]

**Pull request merged**